### PR TITLE
Time integration refactor

### PR DESCRIPTION
**Pull request by Psirus**
We've tried to refactor some of the stuff that has accumulated in `TimeIntegrationBase`. Specifically:

- controlling the time stepping (next step, min/max time step, automatic <-> fixed, etc.) is now a separate class
- all the post-processing has been moved to a separate class
- refactor Newmark; most importantly splitting the gigantic `Solve` routine into several functions

There is still a lot to be done, but too little time, so we would like to merge this, even though there is still a lot of "unattractive" areas in time integration. However, we're open to any comments and suggestions you may have on improving our attempt.

When reviewing, don't be intimidated by the size of the PR. A lot of this is just API changes. Take a look at `TimeIntegrationBase`, `NewmarkDirect`, `PostProcessor` and `TimeControl`.

### CODE REVIEW

File: src/mechanics/timeIntegration/RungeKuttaBase.h [hunk 1]
```diff
@@ -38,7 +38,22 @@ class RungeKuttaBase : public TimeIntegrationBase
...
     virtual bool HasTimeChanged(int rStage) const = 0;
 
+
+    // temporary fix
+    //! @brief sets the  time step for the time integration procedure (initial value)
+    virtual void SetTimeStep(double rTimeStep) override
+    {
+        mTimeStep = rTimeStep;
+    }
+
```
Comment (phuschke) [inline src/mechanics/timeIntegration/RungeKuttaBase.h]: What is the _temporary fix_ for? Why did mTime and mTimeStep moved from TimeIntegrationBase?

File: src/mechanics/timeIntegration/TimeControl.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+
+
+void NuTo::TimeControl::ScaleTimeStep(double scaleFactor)
+{
+    if (scaleFactor<= 0.0)
+        throw MechanicsException(__PRETTY_FUNCTION__,"Scaling factor must be a positive number");
+
+    mTimeStepScaleFactor *= scaleFactor;
+    UpdateTimeStep(-1.,-1.,true);
```
Comment (phuschke) [inline src/mechanics/timeIntegration/TimeControl.cpp]: Why are iterations and maxIterations doubles? What does maxIteration=-1. mean?
`UpdateTimesStep(double iterations, double maxIterations, bool)`

File: src/mechanics/timeIntegration/TimeIntegrationBase.h [hunk 3]
```diff
@@ -195,6 +122,14 @@ class TimeIntegrationBase
...
     {
         mAutomaticTimeStepping = rAutomaticTimeStepping;
+        if(mAutomaticTimeStepping)
+        {
+            mTimeControl.UseDefaultAutomaticTimestepping();
+        }
+        else
+        {
+
+        }
```
Comment (phuschke) [inline src/mechanics/timeIntegration/TimeIntegrationBase.h]: Is something missing here?

File: src/mechanics/timeIntegration/RungeKuttaBase.h [hunk 4]
```diff
@@ -38,7 +38,22 @@ class RungeKuttaBase : public TimeIntegrationBase
     //! @brief ... return, if time (e.g. for the calculation of external loads) has changed
     virtual bool HasTimeChanged(int rStage) const = 0;
 
+
+    // temporary fix
+    //! @brief sets the  time step for the time integration procedure (initial value)
+    virtual void SetTimeStep(double rTimeStep) override
```
Comment (phuschke) [inline src/mechanics/timeIntegration/RungeKuttaBase.h:44]: What is the temporary fix for? Why did mTime and mTimeStep moved from TimeIntegrationBase?

**Review by phuschke — COMMENTED** [review]
Good work. Thank you for improving and cleaning the time integration,

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 5]
```diff
@@ -1,499 +1,182 @@
...
-void NuTo::NewmarkDirect::Info() const
+void NewmarkDirect::Solve(double rTimeDelta)
 {
-    NewmarkBase::Info();
-}
-
-void NuTo::NewmarkDirect::Solve(double rTimeDelta)
-{
-    NuTo::Timer timerFull(__FUNCTION__, mStructure->GetShowTime(), mStructure->GetLogger());
+    mTimeControl.SetTimeFinal(rTimeDelta);
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp]: Assume there was a previous solve up to say t=10, and then there is a new solve for another 10 time units. What is rTimeDelta and Final time (10 or 20?).

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 6]
```diff
@@ -1,499 +1,182 @@
...
-    evaluate_Hessian0_Hessian1[eStructureOutput::HESSIAN0] = &hessian0;
-
-    if (mStructure->GetNumTimeDerivatives() >= 1 && mMuDampingMass == 0.)
+    while (not mTimeControl.Finished())
     {
-        hessian1.Resize(dofStatus.GetNumActiveDofsMap(), dofStatus.GetNumDependentDofsMap());
-        evaluate_InternalGradient_Hessian0Hessian1[eStructureOutput::HESSIAN1] = &hessian1;
-        evaluate_Hessian0_Hessian1[eStructureOutput::HESSIAN1] = &hessian1;
-    }
+        mTimeControl.Proceed();
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp:44]: What exactly does proceed? I think the name is not very intuitive.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 7]
```diff
@@ -1,499 +1,182 @@
...
-                                            << "] Trial residual: " << trialNormResidual << "\n";
+            const auto intForce = EvaluateInternalGradient();
 
-                    alpha *= 0.5;
+            structureResidual = CalculateResidual(intForce, extForce, hessians[2], trial_dof_dt1, trial_dof_dt2);
+            residual = Assembler::ApplyCMatrix(structureResidual, constraintMatrix);
 
-                } while (mPerformLineSearch && alpha > mMinLineSearchStep &&
-                         trialNormResidual > (1. - alpha) * normResidual);
+            trialNormResidual = residual.CalculateInfNorm();
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp]: Is the InfNorm the standard to check convergence? IMO this would mean that by reducing the mesh size the acceptable error in the solution increases.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 8]
```diff
@@ -1,499 +1,182 @@
...
-        //! to work fine, but valgrind tells you otherwise. Problem is, that the DofType is changed during the function
-        //! call inside
-        //! the loop which leads to reads in already freed blocks of memory
-        //! !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
-        std::set<Node::eDof> currentDofSet = dofStatus.GetDofTypes();
+        // LEAVE IT OR FIX IT IF YOU CAN:
+        // If you dont make a copy of the dof set with dofStatus.GetDofTypes() and use it directly in the for loop
+        // everything seems to work fine, but valgrind tells you otherwise. Problem is, that the DofType is changed
+        // during the function call inside the loop which leads to reads in already freed blocks of memory
+        std::set<Node::eDof> currentDofSet = mStructure->GetDofStatus().GetDofTypes();
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp:50]: Why do we change the doftype during the function call?

File: src/mechanics/timeIntegration/TimeControl.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
+#include "TimeControl.h"
+
+
+
+void NuTo::TimeControl::Proceed()
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/TimeControl.cpp]: The name does not stand for itself. Maybe something like ProceedToNextTimeStep

File: src/mechanics/timeIntegration/TimeControl.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+void NuTo::TimeControl::UseEquidistantTimestepping()
+{
+    SetTimeStepFunction([](TimeControl& rTimeControl, int iterations, int maxIterations, bool converged)->double
+                        {
+                            return rTimeControl.GetTimeStep();
+                        });
+}
+
+void NuTo::TimeControl::SetMaxTimeStep(double rMaxTimeStep)
```
Comment (joergfunger) [inline src/mechanics/timeIntegration/TimeControl.cpp]: Shouldn't the max and min time step not be part of this TimeStepFunction? Since they only make sense for adaptive time stepping, whereas for constant steps this is not necessary?

**Review by joergfunger — APPROVED** [review]
Thanks for restructuring the time integration. The tmp changes to the explicit time integration schemes should have some priority for the next NuTo session.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 11]
```diff
@@ -1,499 +1,182 @@
...
-            PrintInfoStagger();
 
-            // ******************************************************
-            mStructure->Evaluate(inputMap, evaluate_Hessian0_Hessian1);
-            // ******************************************************
+std::pair<int, BlockScalar> NewmarkDirect::FindEquilibrium(StructureOutputBlockVector& structureResidual,
+                                                           const StructureOutputBlockVector& extForce,
+                                                           StructureOutputBlockVector& delta_dof_dt0,
+                                                           std::array<StructureOutputBlockVector, 3>& dof_dt,
+                                                           const BlockSparseMatrix& constraintMatrix)
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp]: I would remove constraintMatrix from the argument list, zero-cost-getter mStructure->GetConstraintMatrix() (or something similar)

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 12]
```diff
@@ -584,84 +248,320 @@ void NuTo::NewmarkDirect::PrintInfoStagger() const
...
+    auto inputMap = CreateInputMap();
+    mStructure->Evaluate(inputMap, evalHessians);
+    if (mCheckCoefficientMatrix)
+        mStructure->ElementCheckHessian0(1.e-6, 1.e-8);
+
+    return hessians;
+}
+
+
+std::pair<StructureOutputBlockVector, std::array<StructureOutputBlockMatrix, 3>>
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp:387]: How about `using Hessians = std::array<StructureOutputBlockMatrix, 3>`?

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 13]
```diff
@@ -584,84 +248,320 @@ void NuTo::NewmarkDirect::PrintInfoStagger() const
...
+NuTo::NewmarkDirect::EvaluateGradientAndHessians()
+{
+    const auto& dofStatus = mStructure->GetDofStatus();
+    StructureOutputBlockVector intForce(dofStatus, true);
+    std::array<StructureOutputBlockMatrix, 3> hessians = {StructureOutputBlockMatrix(dofStatus, true),
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp:392]: We could call `EvaluateHessians()` for the hessians part and `EvaluateGradient()` for the gradient. This means that they are not evaluated simultaneously, causing some overhead. IMO this overhead is small compared to `Solve()`.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp [hunk 14]
```diff
@@ -584,84 +248,320 @@ void NuTo::NewmarkDirect::PrintInfoStagger() const
...
+                                                        StructureOutputBlockVector(dofStatus, true),
+                                                        StructureOutputBlockVector(dofStatus, true)};
+    StructureOutputBlockVector delta_dof_dt0(dofStatus, true);
+    const auto& constraintMatrix = mStructure->GetAssembler().GetConstraintMatrix();
+
+    // Needed for mTimeControl.Proceed() function
+    int timeStepMaxIterations = 0;
+    bool converged = true;
+
+    for (const auto& activeDofs : mStepActiveDofs)
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/NewmarkDirect.cpp:433]: This is quite a big method and may need further splits. Maybe into something like
~~~cpp
auto something = TrialState();
auto somethingElse = EquilibriumIterations();
if (converged)
   HandleConvergedSolution();
else
  ...
~~~

File: src/mechanics/timeIntegration/TimeControl.cpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+                                                      "the RestorePreviosTime() function of the time control before reducing "
+                                                      "the timestep.");
+}
+
+
+
+void NuTo::TimeControl::SetTimeStep(double timeStep)
+{
+    if (timeStep<=0)
+        throw MechanicsException(__PRETTY_FUNCTION__,"Timestep must be larger tham 0!");
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/TimeControl.cpp]: than :)

File: src/mechanics/timeIntegration/TimeControl.h [hunk 16]
```diff
@@ -0,0 +1,137 @@
+#pragma once
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/TimeControl.h:1]: This file (among others) needs a clang format run.

File: src/mechanics/timeIntegration/postProcessing/PostProcessor.cpp [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+#include "mechanics/timeIntegration/postProcessing/ResultNodeAcceleration.h"
+#include "mechanics/timeIntegration/postProcessing/ResultGroupNodeForce.h"
+#include "mechanics/timeIntegration/postProcessing/ResultElementIpData.h"
+
+using namespace NuTo;
+
+
+void PostProcessor::PostProcess(const StructureOutputBlockVector& outOfBalance)
+{
+    if (mResultDir.length() == 0)
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/postProcessing/PostProcessor.cpp]: if (mResultDir.empty()) is better IMO.

File: src/mechanics/timeIntegration/postProcessing/PostProcessor.h [hunk 18]
```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <string>
+#include <boost/ptr_container/ptr_vector.hpp>
+
+#include "mechanics/structures/StructureBase.h"
+#include "mechanics/timeIntegration/TimeControl.h"
+#include "mechanics/timeIntegration/postProcessing/ResultBase.h"
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/postProcessing/PostProcessor.h]: All those includes can be forwarded, right?
(StructureBase, TimeControl, ResultBase)

File: src/mechanics/timeIntegration/postProcessing/ResultElementIpData.cpp [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+        break;
+    case IpData::eIpStaticDataType::DAMAGE:
+        numComponents = 1;
+        break;
+    case IpData::eIpStaticDataType::SLIP:
+    case IpData::eIpStaticDataType::BOND_STRESS:
+        numComponents = 3;
+        break;
+    default:
+        throw MechanicsException(std::string(__PRETTY_FUNCTION__) + "\t: Ip data type not supported yet.");
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/postProcessing/ResultElementIpData.cpp]: Since we are touching it right now, this exception is equivalent to
`throw MechanicsException(__PRETTY_FUNCTION__,  "Ip data type not supported yet.");`

File: test/mechanics/timeIntegration/TimeControl.cpp [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+    std::string expectedMsg = "";
+    std::function<bool(MechanicsException const&)> CheckException
+            = [&expectedMsg](MechanicsException const& ex)->bool
+                {
+                    return CheckExceptionMessageCorrect(ex,expectedMsg);
+                };
+
+    expectedMsg = "Current timestep is 0 or negative!";
+//    BOOST_CHECK_EXCEPTION(timeControl.SetTimeStepFunction([](double,double)->double{return 0.0;}),
```
Comment (TTitscher) [inline test/mechanics/timeIntegration/TimeControl.cpp]: should be uncommented, right?

File: test/mechanics/timeIntegration/TimeControl.cpp [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "mechanics/timeIntegration/TimeControl.h"
+
+#include <iostream>
+
+using namespace NuTo;
+
+// %%% Helper functions %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+bool CheckExceptionMessageCorrect(MechanicsException const& ex,std::string expectedMsg)
```
Comment (TTitscher) [inline test/mechanics/timeIntegration/TimeControl.cpp]: I think, you should use the BOOST_CHECK_EQUAL macro instead of this method,
something like `BOOST_CHECK_EQUAL(GetRefinedErrorMsg(ex), expectedMsg);`
Or provide further comments on why you did it with a custom implementation.

**Review by TTitscher — CHANGES_REQUESTED** [review]
Thank you for tackling this topic! 
As you said, this is still work in progress, so you do not have to work on all my remarks. But some of them are easy to fix. 
The main reason why I do not approve the merge is clang format. Please run it on all those files.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp (hunk 12 above)
Comment (Psirus) [reply src/mechanics/timeIntegration/NewmarkDirect.cpp:387]: I'm trying to use type aliases sparingly. If I see `std::array<StructureOutputBlockMatrix, 3>` I know what's going on, assuming I know what a `StructureOutputBlockMatrix` is. With `Hessians` I'll have to go look what this new type I've never heard of represents. But I'll bow to the consensus if we want to go that way.

File: test/mechanics/timeIntegration/TimeControl.cpp [hunk 22]
```diff
@@ -38,55 +38,119 @@ BOOST_AUTO_TEST_CASE(Debug_Fuctionality)
...
 
-        timeControl.SetTimeStep(5.5);
+        timeControl.SetTimeStep(5.0);
+        timeControl.UseEquidistantTimestepping();
 
         timeControl.Proceed();
         timeControl.Proceed();
-        timeControl.Proceed();
-        BOOST_CHECK_EQUAL(timeControl.GetCurrentTime(), 19.5);
+        BOOST_CHECK_EQUAL(timeControl.GetCurrentTime(), 20.0);
```
Comment (TTitscher) [inline test/mechanics/timeIntegration/TimeControl.cpp]: Please use `BOOST_CHECK_CLOSE(a, b, tolerance)` for floating point comparisons.

File: src/mechanics/timeIntegration/RungeKuttaBase.h (hunk 4 above)
Comment (vhirtham) [reply src/mechanics/timeIntegration/RungeKuttaBase.h:44]: Because all time related stuff is now done by the TimeControl class. But it is not that simple to replace them in all derived time integration schemes. Plan is to that step by step.

File: src/mechanics/timeIntegration/NewmarkDirect.cpp (hunk 5 above)
Comment (vhirtham) [reply src/mechanics/timeIntegration/NewmarkDirect.cpp]: Well rTimeDelta is somehow misleading. It should be renamed to rTimeFinal or something like this. We just didn't change it. So you provide the time where you like to stop when you call solve.

File: test/mechanics/timeIntegration/TimeControl.cpp (hunk 21 above)
Comment (vhirtham) [reply test/mechanics/timeIntegration/TimeControl.cpp]: This is just a workaround. Just wait what we decide concerning the exception issue

File: test/mechanics/timeIntegration/TimeControl.cpp (hunk 22 above)
Comment (vhirtham) [reply test/mechanics/timeIntegration/TimeControl.cpp]: Dont know about all the boost test features yet. Will change it if  I don't forget to. :p

File: test/mechanics/timeIntegration/TimeControl.cpp [hunk 23]
```diff
@@ -38,55 +38,119 @@ BOOST_AUTO_TEST_CASE(Debug_Fuctionality)
                 };
 
     expectedMsg = "Current timestep is 0 or negative!";
-//    BOOST_CHECK_EXCEPTION(timeControl.SetTimeStepFunction([](double,double)->double{return 0.0;}),
-//                          MechanicsException,CheckException);
-
-
-
+    timeControl.SetTimeStepFunction([](TimeControl&,double,double,bool)->double{return 0.0;});
```
Comment (TTitscher) [inline test/mechanics/timeIntegration/TimeControl.cpp]: The method is defined to take
`std::function<double(TimeControl&, int, int, bool)> timestepFunction`
Since the test is some kind of reference for the usage of the class, this should be correct here (even though it has no influence in this case...).

**Comment by TTitscher** [comment]
Very nice that you checked all the edge cases! But IMO this is a bit over the top and a lot of duplicate code (the actual exception messages). Just checking that an exception is thrown should be enough. Your debugger will tell you what and why.
What do the others think?

File: test/mechanics/timeIntegration/TimeControl.cpp (hunk 23 above)
Comment (vhirtham) [reply test/mechanics/timeIntegration/TimeControl.cpp]: Yeah you are right. Problem is, that was the old function interface and I changed it due to a hint of pHuschke. I missed that one. Strange though, that the compiler accepts it.

File: src/mechanics/timeIntegration/TimeControl.cpp [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+    mTimeStep = timeStep;
+}
+
+void NuTo::TimeControl::SetTimeFinal(double timeFinal)
+{
+    if (timeFinal <= mCurrentTime)
+    {
+        throw MechanicsException(__PRETTY_FUNCTION__,"Final time must be larger than current time!");
+    }
+    mTimeFinal          = timeFinal;
```
Comment (TTitscher) [inline src/mechanics/timeIntegration/TimeControl.cpp]: Way too many spaces here... Please setup clang-format and run it! Merci much! (Then, IMO, this can be merged)

**Comment by vhirtham** [comment]
Well concerning the exception message checking. That's related to the exception issue. It will be changed when we use specific exception types to distinguish between them.  I want to test specific exceptions because I want to make sure, that the "right" exception is triggered in the "right" circumstances. If you change something and accidentally mess up the exception hierarchy the "right" one might never be trigger because another one is thrown before. 
This one exception might give you a totally wrong hint whats wrong. As a stupid example: You get a "No Convergence" exception because you divided somewhere on the way by 0 and the check for that ends up to be somewhere after the convergence check. As I said, stupid example, but I already got those cases, where the meaningful exception was shadowed by some other, which took me a lot of time to figure out.  
So in the end: I don't think that everybody should test every edge case, but we should probably not complain about to much being tested. Who cares about some code repetition in a rather small test file, that nobody touches anymore, except when it fails?

File: src/mechanics/timeIntegration/TimeIntegrationBase.h (hunk 3 above)
Comment (vhirtham) [reply src/mechanics/timeIntegration/TimeIntegrationBase.h]: Yep. Fixed it

File: src/mechanics/timeIntegration/NewmarkDirect.cpp (hunk 5 above)
Comment (vhirtham) [reply src/mechanics/timeIntegration/NewmarkDirect.cpp]: changed parameter name to timeFinal and cahnged the doxy-doc in the TimeIntegrationBase and NewmarkDirect headers

**Review by TTitscher — APPROVED** [review]
